kernel/src/hil/display.rs

```diff
@@ -0,0 +1,376 @@
+    /// If the screen supports such a feature, the driver should request this information
+    /// from the screen upfront.
+    fn get_num_supported_modes(&self) -> usize;
+
+    /// Can be called with an index from 0 .. count-1 and will
+    /// a tuple (width, height) with the current resolution (in pixels).
+    /// note that width and height may change due to rotation
+    ///
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_supported_mode(&self, index: usize) -> Option<GraphicsMode>;
+}
+
+/* Screen */
+pub trait Screen<'a> {
+    /// Returns the current rotation.
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_rotation(&self) -> Rotation;
+
+    /// Sets the rotation of the display.
+    /// The function should return Ok(()) if the request is registered
+    /// and will be sent to the screen.
+    /// Upon Ok(()), the caller has to wait for the `command_complete` callback function
+    /// that will return the actual Result<(), ErrorCode> after setting the rotation.
+    ///
+    /// Note that in the case of `Rotated90` or `Rotated270`, this will swap the width and height.
+    fn set_rotation(&self, rotation: Rotation) -> Result<(), ErrorCode>;
+
+    /// Controls the screen power supply.
+    ///
+    /// Use it to initialize the display device.
+    ///
+    /// For screens where display needs nonzero brightness (e.g. LED),
+    /// this shall set brightness to a default value
+    /// if `set_brightness` was not called first.
+    ///
+    /// The device may implement power independently from brightness,
+    /// so call `set_brightness` to turn on/off the module completely.
+    ///
+    /// When finished, calls `ScreenClient::screen_is_ready`,
+    /// both when power was enabled and disabled.
+    fn set_power(&self, enabled: bool) -> Result<(), ErrorCode>;
+
+    /// Inverts the colors.
+    fn invert_on(&self) -> Result<(), ErrorCode>;
+
+    /// Reverts the colors to normal.
+    fn invert_off(&self) -> Result<(), ErrorCode>;
+
+    /// Sets the display brightness value
+    ///
+    /// Depending on the display, this may not cause any actual changes
+    /// until and unless power is enabled (see `set_power`).
+    ///
+    /// The following values must be supported:
+    /// - 0 - completely no light emitted
+    /// - 1..MAX_BRIGHTNESS - on, set brightness to the given level
+    ///
+    /// The display should interpret the brightness value as *lightness*
+    /// (each increment should change preceived brightness the same).
+    /// 1 shall be the minimum supported brightness,
+    /// `MAX_BRIGHTNESS` and greater represent the maximum.
+    /// Values in between should approximate the intermediate values;
+    /// minimum and maximum included (e.g. when there is only 1 level).
+    fn set_brightness(&self, brightness: usize) -> Result<(), ErrorCode>;
```

Comment:
I don't think brightness should be in the same trait as inversion. One is very important, the other very unimportant.

Reply:
I agree they might not be in the best place. Where would you see the invert functions?

Reply:
In https://github.com/tock/tock/pull/3011/ I took them out into a separate trait with extra builtin functionality, together with setting rotation. Either way, it should not share the trait with power and brightness control, because those are expected out of almost all displays.